### DISCUSSION TITLE
feat(breaking): update startWith so that it calls a function to get first value

### DISCUSSION THREAD

**Discussion by jorroll** [ReactiveX/rxjs#6235] [Ideas / Feature request]
# Feature Request

_This is a re-opening of issue #5156 which was originally closed because, at the time, it was stated that breaking changes would not be considered to the RXJS core library. Seeing as RXJS has a breaking release being worked on (v7), I'm resurfacing this old discussion._

This is a request to change the `startWith` operator so it expects a function which is called to retrieve the first value.

I.e. update `startWith` interface from this:

```ts
startWith<T, D>(...array: Array<D | SchedulerLike>): OperatorFunction<T, T | D>
```

to this:

```ts
startWith<T, D>(...array: Array<(() => D) | SchedulerLike>): OperatorFunction<T, T | D>
```

See issue #5156 for previous discussion on this topic including [this closing comment](https://github.com/ReactiveX/rxjs/issues/5156#issuecomment-557349717) by `@kwonoj`.

> Closing as the general intention has explained and we will not make changes into core at this moment.

**Comment by kwonoj** [comment]
Moved into discussions since issue is being used for core team's bug tracking purpose.

**Comment by cartant** [comment]
A question related to this has popped up before. There is already a general solution in the library for making things lazy: `defer`. See [this comment](https://github.com/ReactiveX/rxjs/discussions/6082#discussioncomment-434667). Which is basically the [same comment](https://github.com/ReactiveX/rxjs/issues/5156#issuecomment-557346452) that I made in the original issue.

**Reply by jorroll** [reply]
Yes, you suggested this workaround in [#5156](https://github.com/ReactiveX/rxjs/issues/5156#issuecomment-557346452) and it is still the current workaround I use (and a much appreciated suggestion). While the ease of a workaround greatly reduces the need for improving the `startWith` api, I, personally, don't think it eliminates the need to improve the api. Given that breaking changes are already happening in v7, I resurfaced this discussion in the hope that an improved `startWith` can be included in v7. Obviously I understand if the maintainers feel differently tho.

**Comment by benlesh** [comment]
you can always use `concat`:

```ts
concat(
  defer(() => { /* get value here */ return of('value'); }),
  source
)
```

Or roll your own operator:

```ts
const startWithDeferred = <T, V>(factory: () => V) => (source: Observable<T>) => concat(defer(() => of(factory())), source);
```
